distsys/resources/tcpmailboxes.go

```diff
@@ -0,0 +1,350 @@
+			}
+		}()
+
+		return &tcpMailboxesLocalArchetypeResource{
+			msgChannel: msgChannel,
+		}
+	})
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Abort() chan struct{} {
+	res.readBacklog = append(res.readsInProgress, res.readBacklog...)
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) PreCommit() chan error {
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Commit() chan struct{} {
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) ReadValue() (distsys.TLAValue, error) {
+	// if a critical section previously aborted, already-read values will be here
+	if len(res.readBacklog) > 0 {
+		value := res.readBacklog[0]
+		res.readBacklog[0] = distsys.TLAValue{} // ensure this TLAValue is null, otherwise it will dangle and prevent potential GC
+		res.readBacklog = res.readBacklog[1:]
+		res.readsInProgress = append(res.readsInProgress, value)
+		return value, nil
+	}
+
+	// otherwise, either pull a notification + atomically read a value from the buffer, or time out
+	select {
+	case msg := <-res.msgChannel:
+		res.readsInProgress = append(res.readsInProgress, msg)
+		return msg, nil
+	case <-time.After(tcpMailboxesSendTimeout):
+		return distsys.TLAValue{}, distsys.ErrCriticalSectionAborted
+	}
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) WriteValue(value distsys.TLAValue) error {
+	panic(fmt.Errorf("attempted to write value %v to a local mailbox archetype resource", value))
+}
+
+type tcpMailboxesRemoteArchetypeResource struct {
+	distsys.ArchetypeResourceLeafMixin
+	dialAddr string
+
+	inCriticalSection bool
+	conn              net.Conn
+	connEncoder       *gob.Encoder
+	connDecoder       *gob.Decoder
+}
+
+var _ distsys.ArchetypeResource = &tcpMailboxesRemoteArchetypeResource{}
+
+func tcpMailboxesRemoteArchetypeResourceMaker(dialAddr string) distsys.ArchetypeResourceMaker {
+	return distsys.ArchetypeResourceMakerFn(func() distsys.ArchetypeResource {
+		return &tcpMailboxesRemoteArchetypeResource{
+			dialAddr: dialAddr,
+		}
+	})
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) ensureConnection() error {
+	if res.conn == nil {
+		var err error
+		res.conn, err = net.Dial("tcp", res.dialAddr)
+		if err != nil {
+			res.conn, res.connEncoder, res.connDecoder = nil, nil, nil
+			fmt.Printf("failed to dial %s, aborting after 50ms: %v", res.dialAddr, err)
+			time.Sleep(tcpMailboxesConnectionDroppedRetryDelay)
+			return distsys.ErrCriticalSectionAborted
+		}
+		res.connEncoder = gob.NewEncoder(res.conn)
+		res.connDecoder = gob.NewDecoder(res.conn)
+	}
+	return nil
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) Abort() chan struct{} {
+	// nothing to do; the remote end tolerates just starting over with no explanation
+	res.inCriticalSection = false // but note to ourselves that we are starting over, so we re-send the begin record
+	return nil
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) PreCommit() chan error {
```

Comment:
Note that this function can be blocking forever.

Reply:
Is this because I didn't set a timeout, like the other comment below? I forgot to think about that...

Reply:
Yeah, it's same as `WriteValue`.